future-apps/payroll/contracts/Payroll.sol

```diff
@@ -634,4 +667,79 @@ contract Payroll is EtherTokenConstant, IForwarder, IsContract, AragonApp {
 
         return xrt;
     }
+
+    /**
+     * @dev Loops over tokens to send requested amount to the employee
+     * @param _employeeId Employee's identifier
+     * @param _totalAmount Total amount to be transferred to the employee distributed through the setup tokens allocation.
+     * @param _reference String detailing payment reason.
+     * @return True if there was at least one token transfer.
+     */
+    function _transferTokensAmount(uint256 _employeeId, uint256 _totalAmount, string _reference) private returns (bool somethingPaid) {
+        Employee storage employee = employees[_employeeId];
+        for (uint256 i = 0; i < allowedTokensArray.length; i++) {
+            address token = allowedTokensArray[i];
+            if (employee.allocation[token] != uint256(0)) {
+                uint128 exchangeRate = _getExchangeRate(token);
+                require(exchangeRate > 0, ERROR_EXCHANGE_RATE_ZERO);
+                // Salary converted to token and applied allocation percentage
+                uint256 tokenAmount = _totalAmount.mul(exchangeRate).mul(employee.allocation[token]);
+                // Divide by 100 for the allocation and by ONE for the exchange rate
+                tokenAmount = tokenAmount / (100 * ONE);
+                finance.newPayment(token, employee.accountAddress, tokenAmount, 0, 0, 1, _reference);
+                emit SendPayment(employee.accountAddress, token, tokenAmount, _reference);
+                somethingPaid = true;
+            }
+        }
+    }
```

Comment:
⚠️ We should add a reentrancy guard to this function and `Payroll.determineAllocation()` (or perhaps all the functions with the `employeeMatches` modifier).

Right now, with ETH, you can re-enter into `Payroll.determineAllocation()` and modify your allocation so that a later token becomes 100% of your allocation. The worse case of this is if you set ETH to 100%, and then upon paying ETH, your contract re-enters into `Payroll.determineAllocation()` and sets a later token to be 100%.

Reply:
Will move this one to a separate issue